File: C++/cellmap.cpp
#include "cellmap.h"
#include "utils.hpp"

void Cellmap::init(int w, int h, int cs) {
    GW = w;
    GH = h;
    CS = cs;
    old_buf = 0;
    curr_buf = 1;
    cell_color_alive = sf::Color(255, 128, 0);
    cell_color_dead = sf::Color(32, 32, 32);

    init_cells();
    randomize_cells();
}

void Cellmap::update() {
    if (!paused)
        pass_generation();
}

void Cellmap::render(sf::RenderWindow* window) {
    window->draw(quads);
}

void Cellmap::toggle_pause() {
    paused = !paused;
}

void Cellmap::init_cells() {
    // init buffers
    for (int buffers = 0; buffers < 2; buffers++) {
        vector< vector<uint8_t> > outer;
        for (int j = 0; j < GH; j++) {
            vector<uint8_t> inner;
            for (int i = 0; i < GW; i++) {
                inner.push_back(0);
            }
            outer.push_back(inner);
        }
        cells.push_back(outer);
    }

    // init vertexarray
    quads.setPrimitiveType(sf::Quads);
    quads.resize(GW * GH * 4);

    if (CS < 4) pad = 0; // otherwise if cells are 2x2 px, padding will make them 0x0 px

    for (int j = 0; j < GH; j++) {
        for (int i = 0; i < GW; i++) {
            int INDEX = (i+j*GW) * 4;
            sf::Vertex* quad = &quads[ INDEX ];
            quad[0].position = sf::Vector2f(  i   *CS +pad,    j   *CS +pad );    // top left
            quad[1].position = sf::Vector2f( (i+1)*CS -pad,    j   *CS +pad );    // top right
            quad[2].position = sf::Vector2f( (i+1)*CS -pad,   (j+1)*CS -pad );    // bot right
            quad[3].position = sf::Vector2f(  i   *CS +pad,   (j+1)*CS -pad );    // bot left

            update_quad(quad, cell_color_dead);
        }
    }
}

void Cellmap::update_quad(sf::Vertex* quad, sf::Color c) {
    quad[0].color = c;
    quad[1].color = c;
    quad[2].color = c;
    quad[3].color = c;
}

void Cellmap::randomize_cells() {
    for (int j = 0; j < GH; j++) {
        for (int i = 0; i < GW; i++) {
            bool r = utils::irand(0, 100) > 50;
            if (r) {
                cells[curr_buf][j][i] = 1;
                update_quad(&quads[(i+j*GW) * 4], cell_color_alive);
            } else {
                cells[old_buf][j][i] = 0;
                update_quad(&quads[(i+j*GW) * 4], cell_color_dead);
            }
        }
    }
}

void Cellmap::pass_generation() {
    old_buf = 1 - old_buf;
    curr_buf = 1 - curr_buf;

    for (int j = 0; j < GH; j++) {
        for (int i = 0; i < GW; i++) {
            // wrap around
            int j_ = j-1 >=  0   ?   j-1   :  GH-1;
            int _j = j+1 <  GH   ?   j+1   :     0;
            int i_ = i-1 >=  0   ?   i-1   :  GW-1;
            int _i = i+1 <  GW   ?   i+1   :     0;

            int n =   cells[old_buf][ j_    ][ i_    ]
                    + cells[old_buf][ j_    ][   i   ]
                    + cells[old_buf][ j_    ][    _i ]
                    + cells[old_buf][   j   ][ i_    ]
                    + cells[old_buf][   j   ][    _i ]
                    + cells[old_buf][    _j ][ i_    ]
                    + cells[old_buf][    _j ][   i   ]
                    + cells[old_buf][    _j ][    _i ];

            bool alive = n == 3 || ( cells[old_buf][j][i] && n == 2);
            cells[curr_buf][j][i] = (int)alive;

            if (alive)  update_quad(&quads[(i+j*GW) * 4], cell_color_alive);
            else        update_quad(&quads[(i+j*GW) * 4], cell_color_dead);

        }
    }
}

File: Lua/cellmap.lua

function Cellmap(w, h, cs)
	local t = {
		CS = cs,
		GW = w,
		GH = h,
		pad = cs > 2 and 1 or 0, -- otherwise if cells are 2x2 px, padding will make them 0x0 px
		old_buf = 0,
		curr_buf = 1,
		cells = {},
		cell_color_alive = {1, 0.5, 0, 1},
		cell_color_dead = {0.1, 0.1, 0.1, 1},
	}

	function t.init()
		t.init_cells()
		t.randomize_cells()
	end

	function t.init_cells()
		for bufs=0, 1 do
			t.cells[bufs] = {}
			for j=0, t.GH-1 do
				t.cells[bufs][j] = {}
				for i=0, t.GW-1 do
					t.cells[bufs][j][i] = 0
				end
			end
		end
	end

	function t.randomize_cells()
		for j=0, t.GH-1 do
			for i=0, t.GW-1 do
				local alive = love.math.random()*100 < 50
				if alive == true then	t.cells[t.curr_buf][j][i] = 1
				else 					t.cells[t.curr_buf][j][i] = 0
				end
			end
		end
	end

	function t.draw()
		local size = t.CS-t.pad*2
		for j=0, t.GH-1 do
			for i=0, t.GW-1 do
				local x = i*t.CS + t.pad
				local y = j*t.CS + t.pad
				if t.cells[t.curr_buf][j][i] == 1 then
					love.graphics.setColor(t.cell_color_alive)
				else
					love.graphics.setColor(t.cell_color_dead)
				end
				love.graphics.rectangle('fill', x, y, size, size)
			end
		end
	end

	function t.update(dt)
		t.next_gen()
	end

	function t.next_gen()
		t.curr_buf = 1 - t.curr_buf
		t.old_buf = 1 - t.old_buf

		for j=0, t.GH-1 do
			for i=0, t.GW-1 do
				local j_, _j, i_, _i =
				j-1 <      0 and t.GH-1 or j-1,
				j+1 > t.GH-1 and      0 or j+1,
				i-1 <      0 and t.GW-1 or i-1,
				i+1 > t.GW-1 and      0 or i+1

				local n = t.cells[t.old_buf][ j_    ][ i_    ]	-- this is the slow part
						+ t.cells[t.old_buf][ j_    ][   i   ]
						+ t.cells[t.old_buf][ j_    ][    _i ]
						+ t.cells[t.old_buf][   j   ][ i_    ]
						+ t.cells[t.old_buf][   j   ][    _i ]
						+ t.cells[t.old_buf][    _j ][ i_    ]
						+ t.cells[t.old_buf][    _j ][   i   ]
						+ t.cells[t.old_buf][    _j ][    _i ]

				t.cells[t.curr_buf][j][i] = (n==3 or (n==2 and t.cells[t.old_buf][j][i] == 1)) and 1 or 0
			end
		end
	end

	return t
end